[0.9.5 nightly-ff997a7] Dropping database under a write load causes panics

While running a test, I dropped the database that the test was writing to. This resulted in the following panic.

```
[tsm1] 2015/10/21 16:09:15 rewriting index file /var/opt/influxdb/data/stress/default/14/0001170.tsm1 with /var/opt/influxdb/data/stress/default/14/0001173.tsm1
[query] 2015/10/21 16:09:17 DROP DATABASE stress
[tsm1wal] 2015/10/21 16:09:17 /var/opt/influxdb/data/stress/default/14 flush to index took 2.755019774s
[wal] 2015/10/21 16:09:18 Flush due to idle. Flushing 16 series with 16 points and 740 bytes from partition 1
[wal] 2015/10/21 16:09:18 write to index of partition 1 took 1.430646ms
[monitor] 2015/10/21 16:09:27 failed to store statistics: timeout
[monitor] 2015/10/21 16:09:37 failed to store statistics: timeout
[monitor] 2015/10/21 16:09:47 failed to store statistics: timeout
[query] 2015/10/21 16:09:48 DROP DATABASE stress
[monitor] 2015/10/21 16:09:57 failed to store statistics: timeout
[monitor] 2015/10/21 16:10:07 failed to store statistics: timeout
[tsm1] 2015/10/21 16:10:12 Compaction of /var/opt/influxdb/data/stress/default/14 took 4m13.079764908s
panic: error opening new segment file for wal: open /var/opt/influxdb/data/stress/default/14/_03549.wal: no such file or directory

goroutine 407990 [running]:
github.com/influxdb/influxdb/tsdb/engine/tsm1.(*Log).writeToLog(0xc823cd4700, 0x62101, 0xc84d0dc000, 0x10a15, 0x6218e, 0x0, 0x0)
        /tmp/tmp.vlHqNIFOU1/src/github.com/influxdb/influxdb/tsdb/engine/tsm1/wal.go:411 +0x94b
github.com/influxdb/influxdb/tsdb/engine/tsm1.(*Log).WritePoints(0xc823cd4700, 0xc842c66000, 0x1387, 0x1800, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
        /tmp/tmp.vlHqNIFOU1/src/github.com/influxdb/influxdb/tsdb/engine/tsm1/wal.go:211 +0x35f
github.com/influxdb/influxdb/tsdb/engine/tsm1.(*Engine).WritePoints(0xc823f3d560, 0xc842c66000, 0x1387, 0x1800, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
        /tmp/tmp.vlHqNIFOU1/src/github.com/influxdb/influxdb/tsdb/engine/tsm1/tsm1.go:348 +0x8a
github.com/influxdb/influxdb/tsdb.(*Shard).WritePoints(0xc823f51b80, 0xc842c66000, 0x1387, 0x1800, 0x0, 0x0)
        /tmp/tmp.vlHqNIFOU1/src/github.com/influxdb/influxdb/tsdb/shard.go:263 +0x670
github.com/influxdb/influxdb/tsdb.(*Store).WriteToShard(0xc820086780, 0xe, 0xc842c66000, 0x1387, 0x1800, 0x0, 0x0)
        /tmp/tmp.vlHqNIFOU1/src/github.com/influxdb/influxdb/tsdb/store.go:396 +0x14e
github.com/influxdb/influxdb/cluster.(*PointsWriter).writeToShard.func1(0xc82007a6c0, 0xc830dc56ef, 0x6, 0xc824689580, 0x7, 0xc831be8540, 0x1, 0xe, 0x1, 0xc842c66000, ...)
        /tmp/tmp.vlHqNIFOU1/src/github.com/influxdb/influxdb/cluster/points_writer.go:302 +0x10b
created by github.com/influxdb/influxdb/cluster.(*PointsWriter).writeToShard
        /tmp/tmp.vlHqNIFOU1/src/github.com/influxdb/influxdb/cluster/points_writer.go:334 +0x1c5
```

Verified that returning an error from creating the segments prevents the original panic, but now I'm getting a new panic on the writes below.

```
[tsm1wal] 2015/10/26 10:03:04 addToCache: failed to flush: error creating new wal file: open /Users/jason/.influxdb/data/stress/default/24/_00023.wal: no such file or directory
panic: error writing type to wal: write /Users/jason/.influxdb/data/stress/default/24/_00022.wal: bad file descriptor

goroutine 5747 [running]:
github.com/influxdb/influxdb/tsdb/engine/tsm1.(*Log).writeToLog(0xc82774c200, 0x62101, 0xc83f236000, 0x10a28, 0x62160, 0x0, 0x0)
	/Users/jason/go/src/github.com/influxdb/influxdb/tsdb/engine/tsm1/wal.go:444 +0x2fb
github.com/influxdb/influxdb/tsdb/engine/tsm1.(*Log).WritePoints(0xc82774c200, 0xc825f06000, 0x1387, 0x1800, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
	/Users/jason/go/src/github.com/influxdb/influxdb/tsdb/engine/tsm1/wal.go:215 +0x341
github.com/influxdb/influxdb/tsdb/engine/tsm1.(*Engine).WritePoints(0xc8240a6b40, 0xc825f06000, 0x1387, 0x1800, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0)
	/Users/jason/go/src/github.com/influxdb/influxdb/tsdb/engine/tsm1/tsm1.go:389 +0x8a
github.com/influxdb/influxdb/tsdb.(*Shard).WritePoints(0xc8235ce2c0, 0xc825f06000, 0x1387, 0x1800, 0x0, 0x0)
	/Users/jason/go/src/github.com/influxdb/influxdb/tsdb/shard.go:263 +0x670
github.com/influxdb/influxdb/tsdb.(*Store).WriteToShard(0xc820154000, 0x18, 0xc825f06000, 0x1387, 0x1800, 0x0, 0x0)
	/Users/jason/go/src/github.com/influxdb/influxdb/tsdb/store.go:396 +0x14e
github.com/influxdb/influxdb/cluster.(*PointsWriter).writeToShard.func1(0xc82014a1b0, 0xc8448f8fcf, 0x6, 0xc835449520, 0x7, 0xc83743c420, 0x1, 0x18, 0x1, 0xc825f06000, ...)
	/Users/jason/go/src/github.com/influxdb/influxdb/cluster/points_writer.go:302 +0x10b
created by github.com/influxdb/influxdb/cluster.(*PointsWriter).writeToShard
	/Users/jason/go/src/github.com/influxdb/influxdb/cluster/points_writer.go:334 +0x1c5
```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -76,6 +76,7 @@
 - [#4497](https://github.com/influxdb/influxdb/pull/4497): Fix sequence in meta proto
 - [#3367](https://github.com/influxdb/influxdb/issues/3367): Negative timestamps are parsed correctly by the line protocol.
 - [#4563](https://github.com/influxdb/influxdb/pull/4536): Fix broken subscriptions updates.
+- [#4538](https://github.com/influxdb/influxdb/issues/4538): Dropping database under a write load causes panics
 
 ## v0.9.4 [2015-09-14]
 

--- a/tsdb/engine/tsm1/tsm1.go
+++ b/tsdb/engine/tsm1/tsm1.go
@@ -204,7 +204,11 @@ func (e *Engine) PerformMaintenance() {
 		}
 	}
 
-	go e.Compact(true)
+	go func() {
+		if err := e.Compact(true); err != nil {
+			e.logger.Printf("PerformMaintenance: error during compaction: %v", err)
+		}
+	}()
 }
 
 // Format returns the format type of this engine
@@ -482,7 +486,11 @@ func (e *Engine) Write(pointsByKey map[string]Values, measurementFieldsToSave ma
 	}
 
 	if !e.SkipCompaction && e.shouldCompact() {
-		go e.Compact(false)
+		go func() {
+			if err := e.Compact(false); err != nil {
+				e.logger.Printf("Write: error during compaction: %v", err)
+			}
+		}()
 	}
 
 	return nil

--- a/tsdb/engine/tsm1/tsm1.go
+++ b/tsdb/engine/tsm1/tsm1.go
@@ -325,6 +325,8 @@ func (e *Engine) Close() error {
 	e.filesLock.Lock()
 	defer e.filesLock.Unlock()
 
+	e.WAL.Close()
+
 	// ensure all deletes have been processed
 	e.deletesPending.Wait()
 

--- a/tsdb/engine/tsm1/wal.go
+++ b/tsdb/engine/tsm1/wal.go
@@ -659,9 +659,9 @@ func (l *Log) flush(flush flushType) error {
 			return err
 		}
 		if id <= lastFileID {
-			err := os.Remove(fn)
+			err := os.RemoveAll(fn)
 			if err != nil {
-				return err
+				return fmt.Errorf("failed to remove: %v: %v", fn, err)
 			}
 		}
 	}

--- a/tsdb/engine/tsm1/wal.go
+++ b/tsdb/engine/tsm1/wal.go
@@ -576,6 +576,8 @@ func (l *Log) flush(flush flushType) error {
 	if flush == idleFlush {
 		if l.currentSegmentFile != nil {
 			if err := l.currentSegmentFile.Close(); err != nil {
+				l.cacheLock.Unlock()
+				l.writeLock.Unlock()
 				return err
 			}
 			l.currentSegmentFile = nil
@@ -583,8 +585,9 @@ func (l *Log) flush(flush flushType) error {
 		}
 	} else {
 		if err := l.newSegmentFile(); err != nil {
-			// there's no recovering from this, fail hard
-			panic(fmt.Sprintf("error creating new wal file: %s", err.Error()))
+			l.cacheLock.Unlock()
+			l.writeLock.Unlock()
+			return fmt.Errorf("error creating new wal file: %s", err.Error())
 		}
 	}
 	l.writeLock.Unlock()

--- a/tsdb/engine/tsm1/wal.go
+++ b/tsdb/engine/tsm1/wal.go
@@ -431,8 +431,9 @@ func (l *Log) writeToLog(writeType walEntryType, data []byte) error {
 
 	if l.currentSegmentFile == nil || l.currentSegmentSize > DefaultSegmentSize {
 		if err := l.newSegmentFile(); err != nil {
-			// fail hard since we can't write data
-			panic(fmt.Sprintf("error opening new segment file for wal: %s", err.Error()))
+			// A drop database or RP call could trigger this error if writes were in-flight
+			// when the drop statement executes.
+			return fmt.Errorf("error opening new segment file for wal: %s", err.Error())
 		}
 	}
 

--- a/tsdb/engine/tsm1/wal.go
+++ b/tsdb/engine/tsm1/wal.go
@@ -578,7 +578,7 @@ func (l *Log) flush(flush flushType) error {
 			if err := l.currentSegmentFile.Close(); err != nil {
 				l.cacheLock.Unlock()
 				l.writeLock.Unlock()
-				return err
+				return fmt.Errorf("error closing current segment: %v", err)
 			}
 			l.currentSegmentFile = nil
 			l.currentSegmentSize = 0
@@ -587,7 +587,7 @@ func (l *Log) flush(flush flushType) error {
 		if err := l.newSegmentFile(); err != nil {
 			l.cacheLock.Unlock()
 			l.writeLock.Unlock()
-			return fmt.Errorf("error creating new wal file: %s", err.Error())
+			return fmt.Errorf("error creating new wal file: %v", err)
 		}
 	}
 	l.writeLock.Unlock()

--- a/tsdb/query_executor.go
+++ b/tsdb/query_executor.go
@@ -382,12 +382,17 @@ func (q *QueryExecutor) executeDropDatabaseStatement(stmt *influxql.DropDatabase
 		}
 	}
 
+	// Remove database from meta-store first so that in-flight writes can complete without error, but new ones will
+	// be rejected.
+	res := q.MetaStatementExecutor.ExecuteStatement(stmt)
+
+	// Remove the database from the local store
 	err = q.Store.DeleteDatabase(stmt.Name, shardIDs)
 	if err != nil {
 		return &influxql.Result{Err: err}
 	}
 
-	return q.MetaStatementExecutor.ExecuteStatement(stmt)
+	return res
 }
 
 // executeDropMeasurementStatement removes the measurement and all series data from the local store for the given measurement

--- a/tsdb/store.go
+++ b/tsdb/store.go
@@ -155,7 +155,9 @@ func (s *Store) DeleteDatabase(name string, shardIDs []uint64) error {
 		if shard != nil {
 			shard.Close()
 		}
+		delete(s.shards, id)
 	}
+
 	if err := os.RemoveAll(filepath.Join(s.path, name)); err != nil {
 		return err
 	}

